File: stickyindex/src/main/java/br/com/stickyindex/adapter/IndexAdapter.java
package br.com.stickyindex.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import br.com.stickyindex.R;

/**
 * Created by edgar on 6/4/15.
 */
public class IndexAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private char[] dataSet;
    private RowStyle rowStyle;
    private char prev;

    // CONSTRUCTOR _________________________________________________________________________________
    public IndexAdapter (char[] data, RowStyle style) {
        this.dataSet = data;
        this.rowStyle = style;
    }

    // CALLBACKS ___________________________________________________________________________________
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.sticky_row_details, parent, false);

        if (rowStyle != null) {
            if (rowStyle.getRowHeigh() != -1) {
                android.view.ViewGroup.LayoutParams params = view.getLayoutParams();
                params.height = rowStyle.getRowHeigh().intValue();
                params.width = rowStyle.getStickyWidth().intValue();
                view.setLayoutParams(params);
            }

            TextView index = (TextView) view.findViewById(R.id.sticky_row_index);

            if (rowStyle.getTextColor() != -1) {
                index.setTextColor(rowStyle.getTextColor());
            }

            if (rowStyle.getTextSize().intValue() != -1) {
                index.setTextSize(rowStyle.getTextSize());
            }

            if (rowStyle.getTextStyle() != -1) {
                index.setTypeface(null, rowStyle.getTextStyle());
            }
        }

        return new IndexViewHolder(view);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        IndexViewHolder viewHolder = (IndexViewHolder) holder;


        viewHolder.index.setText(Character.toString(dataSet[position]));

        if (isHeader(position)) {
            viewHolder.index.setVisibility(TextView.VISIBLE);
        } else {
            viewHolder.index.setVisibility(TextView.INVISIBLE);
        }
    }


    // UTIL ________________________________________________________________________________________
    private Boolean isHeader (int pos) {
        if (pos == 0) {
            return Boolean.TRUE;
        } else if (isSameChar(dataSet[pos-1], dataSet[pos])) {
            return Boolean.FALSE;
        } else {
            return Boolean.TRUE;
        }
    }

    private Boolean isSameChar (char prev, char curr) {
        if (Character.toLowerCase(prev) == Character.toLowerCase(curr)) {
            return Boolean.TRUE;
        } else {
            return Boolean.FALSE;
        }
    }

    // GETTERS AND SETTERS _________________________________________________________________________
    @Override
    public int getItemCount() {
        return dataSet.length;
    }

    public void setDataSet(char[] dataSet) {
        this.dataSet = dataSet;
    }

    // ViewHolder class ____________________________________________________________________________
    public static class IndexViewHolder extends RecyclerView.ViewHolder {
        TextView index;

        public IndexViewHolder (View v) {
            super (v);
            index = (TextView) v.findViewById(R.id.sticky_row_index);
        }
    }

    public static class RowStyle {
        Float rowHeigh;
        Float stickyWidth;
        Integer textColor;
        Integer textSize;
        Integer textStyle;

        public RowStyle (Float rHeight, Float sWidth, Integer tColor, Integer tSize, Integer tStyle) {
            rowHeigh = rHeight;
            stickyWidth = sWidth;
            textColor = tColor;
            textSize = tSize;
            textStyle = tStyle;
        }

        public Float getStickyWidth() {
            return stickyWidth;
        }

        public void setStickyWidth(Float stickyWidth) {
            this.stickyWidth = stickyWidth;
        }

        public Float getRowHeigh() {
            return rowHeigh;
        }

        public void setRowHeigh(Float rowHeigh) {
            this.rowHeigh = rowHeigh;
        }

        public Integer getTextColor() {
            return textColor;
        }

        public void setTextColor(Integer textColor) {
            this.textColor = textColor;
        }

        public Integer getTextSize() {
            return textSize;
        }

        public void setTextSize(Integer textSize) {
            this.textSize = textSize;
        }

        public Integer getTextStyle() {
            return textStyle;
        }

        public void setTextStyle(Integer textStyle) {
            this.textStyle = textStyle;
        }
    }
}

File: sticky-index/src/main/java/br/com/stickyindex/adapter/IndexAdapter.kt
package br.com.stickyindex.adapter

import android.support.v7.widget.RecyclerView
import android.util.TypedValue.COMPLEX_UNIT_PX
import android.view.LayoutInflater
import android.view.View
import android.view.View.INVISIBLE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.TextView
import br.com.stickyindex.R

/**
 * Created by edgar on 6/4/15.
 */
class IndexAdapter(
        private var dataSet: CharArray,
        private val rowStyle: RowStyle?
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.sticky_row_details, parent, false)
        applyStyle(view)
        return ViewHolder(view)
    }

    private fun IndexAdapter.applyStyle(view: View) {
        if (rowStyle == null) {
            return
        }

        if (exists(rowStyle.height)) {
            setLayoutParams(view, rowStyle)
        }

        val index = view.findViewById<View>(R.id.sticky_row_index) as TextView

        if (exists(rowStyle.textColor)) {
            index.setTextColor(rowStyle.textColor)
        }

        if (exists(rowStyle.textSize.toInt())) {
            index.setTextSize(COMPLEX_UNIT_PX, rowStyle.textSize)
        }

        if (exists(rowStyle.textStyle)) {
            index.setTypeface(null, rowStyle.textStyle)
        }
    }

    private fun setLayoutParams(view: View, rowStyle: RowStyle) {
        val params = view.layoutParams
        params.height = rowStyle.height.toInt()
        params.width = rowStyle.stickyWidth.toInt()
        view.layoutParams = params
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val viewHolder = holder as ViewHolder
        viewHolder.index.text = Character.toString(dataSet[position])
        if (isHeader(position)!!) {
            viewHolder.index.visibility = VISIBLE
        } else {
            viewHolder.index.visibility = INVISIBLE
        }
    }

    private fun isHeader(position: Int): Boolean? =
            position == 0 || !isSameChar(dataSet[position - 1], dataSet[position])!!

    private fun isSameChar(prev: Char, curr: Char): Boolean? {
        return Character.toLowerCase(prev) == Character.toLowerCase(curr)
    }

    override fun getItemCount(): Int = dataSet.size

    fun setDataSet(dataSet: CharArray) {
        this.dataSet = dataSet
    }

    private fun exists(number: Float): Boolean = number != (-1).toFloat()
    private fun exists(number: Int): Boolean = number != -1

    class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        internal var index: TextView = v.findViewById<View>(R.id.sticky_row_index) as TextView

    }

    data class RowStyle(
            val height: Float,
            val stickyWidth: Float,
            val textColor: Int,
            val textSize: Float,
            val textStyle: Int
    )
}
